gh-63075: IDLE editor - Auto insertion of the closers

wohlganger opened the pull request:
This patch is for Auto insertion of the closing parens, brackets, and braces enhancement. It includes necessary changes to config-extensions.def, and idlelib.test.test_config. It has its own test which covers all non - __init__ statements. As of writing this, idlelib.tests passes on all tests other than calltips, which appears to be a preexisting problem.

<!-- issue-number: bpo-18875 -->
https://bugs.python.org/issue18875
<!-- /issue-number -->


<!-- gh-issue-number: gh-63075 -->
* Issue: gh-63075
<!-- /gh-issue-number -->

wohlganger committed:
parenmatch highlighting options

adds fg color, bg color, font, and underline options for parens highlighting

wohlganger committed:
fixed whitespace issue

wohlganger committed:
only select from highlighting options

wohlganger committed:
fixed whitespace

wohlganger committed:
whitespace

wohlganger committed:
Merge remote-tracking branch 'upstream/master'

wohlganger committed:
#27099 - turn builtin extensions to reg modules

wohlganger committed:
whitespace + fix I thought I already committed.

wohlganger committed:
warning fixes, removed old config entries

wohlganger committed:
calltip append doc fix,  default paren-fore fix, help update

How did the change to the __doc__ of append not get fixed in the tests earlier?

wohlganger committed:
fix calltips

Used incorrect / old python for testing.

wohlganger committed:
removed circular import

wohlganger committed:
fix warnings due to missing default config options in idlelib.config-main.def

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
added format_pmaxw to config-main

wohlganger committed:
Merge remote-tracking branch 'upstream/master'

wohlganger committed:
replaced help.html edit with idle.rst edit

per terryjreedy request.

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
Merge remote-tracking branch 'upstream/master'

wohlganger committed:
added conflict resolved var descriptions to docstring in create_page_highlight.

wohlganger committed:
strip trailing whitespace

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
re-introduced highlight options.

Fixing code conflict ate highlighting options. Needed to be re-introduced.

terryjreedy committed:
News blurb.

wohlganger committed:
Reinstated zoomheight as an extension. Replaced autoexpand with zoomheight as  non-core keybinding test example.

wohlganger committed:
fixed configdialog test (wrong cases), fixed whitespace

wohlganger committed:
update idle.rst: zoomheight is an extension.

wohlganger committed:
more fixes for zoomheight - trying to minimize changes from master

wohlganger committed:
zoomheight config fix, zoomheight should be fully back to the way it was before.

wohlganger committed:
whitespace

wohlganger committed:
fixed force-open completions missing from config-keys and config. Fixed zoomheight configdialog test.

wohlganger committed:
Changes made per terryjreedy request.

Sorry about the formatting; I think that happened when I was trying to fix for the conflicts.

wohlganger committed:
re-added autocomplete wait and format paragraph max width options to general tab

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
changes per terryjreedy

still need to move the imports

wohlganger committed:
fix warnings, errors, bugs

AutoComplete, CodeContext, FormatParagraph, ParenMatch needed special exclusions to have settings read from config-extensions.def . Special exclusions needed for calltips, completions, paren-closed events to prevent them being configurable.

wohlganger committed:
whitespace

wohlganger committed:
bugfix

if the cfg file had one of the old extensions in it, it was showing up in the extensions tab. This fixes that.

terryjreedy committed:
Make changes need to get IDLE to run

terryjreedy committed:
Edit extension/feature files, making existing tests pass.

terryjreedy committed:
Redo feature bindings.

terryjreedy committed:
First draft of dummy extension.

terryjreedy committed:
Fix that eliminate startup error and all but 2 test_idle errors.

terryjreedy committed:
Move widgets, load widgets, add tests, make previous tests pass.

terryjreedy committed:
Add format-paragraph to core keys.

terryjreedy committed:
Merge remote-tracking branch 'origin/master' into pr_2494

terryjreedy committed:
Fix autocomplete and long lines.

terryjreedy committed:
Put fixed-key event-adds in EditorWindow.__init__ for now.

terryjreedy committed:
Don't warn if new core keys not in user key config.

terryjreedy committed:
Make code context inoperative in outwin, shell.

terryjreedy committed:
Change new bindings for MacOsx.

terryjreedy committed:
Disable test dependent on extension Alt keys.

terryjreedy committed:
Minor edits.

terryjreedy committed:
File out news blurb explaining effect for users.

terryjreedy committed:
Don't compute ('') during integer input in entry box.

wohlganger committed:
Merge remote-tracking branch 'upstream/master'

wohlganger committed:
fix conflict

wohlganger committed:
fix conflict

wohlganger committed:
fix conflict, match upstream

wohlganger committed:
match upstream

wohlganger committed:
Auto-insert-parens extension

includes changes to config-extensions.def and test_config.py for it to run and pass tests. New tests for it in test_parenclose. Coverage tool indicates it covers everything but __init__.  All idlelib tests pass, except one for calltips, which I'm fairly sure fails outside of this patch anyway.

wohlganger committed:
removed unused old options

sorry i missed this before uploading

csabella commented on Lib/idlelib/config-extensions.def:
```diff
@@ -63,3 +63,13 @@ z-text= Z
+skip_closures = True
+[ParenClose_cfgBindings]
+p-open = <Key-parenleft> <Key-bracketleft> <Key-braceleft>
+t-open = <Key-'> <Key-">
```
I downloaded the PR to take a look at it and couldn't open IDLE because of this error:

`File "/home/cheryl/cpython/Lib/idlelib/editor.py", line 137, in __init__
    self.apply_bindings()`
  `File "/home/cheryl/cpython/Lib/idlelib/editor.py", line 1071, in apply_bindings
    text.event_add(event, *keylist)`
 ` File "/home/cheryl/cpython/Lib/idlelib/multicall.py", line 374, in event_add
    widget.event_add(self, virtual, seq)`
`  File "/home/cheryl/cpython/Lib/tkinter/__init__.py", line 1653, in event_add
    self.tk.call(args)`
`_tkinter.TclError: bad event type or keysym "'"`

I looked at this page for the keysyms:
https://www.tcl.tk/man/tcl8.4/TkCmd/keysyms.htm

and it has `quotedbl                            34     0x0022` and `quoteright                          39     0x0027`, which seems to fix the error.

Also, `'` and `"` aren't available on the config keys screen, so they may need to be added.

csabella commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,91 @@
+    [ParenClose_cfgBindings]
+    p-open = <Key-parenleft> <Key-bracketleft> <Key-braceleft>
+    t-open = <Key-'> <Key-">
+    p-close = <Key-parenright> <Key-bracketright> <Key-braceright>
```
p-open, t-open, and p-close appear in the keybindings list in Configure IDLE.  Was the intention to reserve these so the keys couldn't be bound to another shortcut?  If so, I think they need to be suppressed from the keybindings selection, but still seen as a duplicate if someone tries to select it.

terryjreedy replied:
If this were implemented as an extension, which it will not be, the section title should be [ParenClose_bindings] because these are fixed, not configurable, bindings.  The user reconfiguration dialog should not allow bare, unmodified <Key-x> bindings, at least not for character keys.  Edit. The section title, which will go away anyway, was changed.  The comment above should have been marked 'Resolved' then.

csabella commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,91 @@
+    symbol is typed and the same one is to the right of it, that symbols is
+    deleted before the new one is typed, effectively skipping over the closure.
+    '''
+    def __init__(self, editwin=None): #setting default to none makes testing easier
```
Inline comments should be written like a sentence - begin with a capital letter and end with a period.

csabella commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,91 @@
+        self.skip_closures = idleConf.GetOption(
+            'extensions', 'ParenClose', 'skip_closures', default=True)
+        #sometimes idleConf loads boolean False as string "False"
+        if self.paren_close == 'False':
```
GetOption has a `type` parameter.  If you set it to `bool`, I think it should always return a boolean.

csabella commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,91 @@
+            else:
+                if self.skip_closures \
+                   and self.text.get(pos, pos + ' +1c') == event.char:
+                    self.text.delete(pos, pos + ' +1c')
```
Although there aren't a lot of lines here, this is duplicate code with `p_close_event` so I'm wondering if a helper function that both would call would be beneficial?

csabella commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,91 @@
+                    self.text.delete(pos, pos + ' +1c')
+                else:
+                    self.text.insert(pos, event.char)
+                    self.text.mark_set('insert', pos)
```
Ditto here with `p_open_event`.  The dictionary could map `'` to `'` and `"` to `"` for it use `closer` instead of `event.char`.  I know the triple quotes are still a special case.

csabella commented:
When I type an opener (parenthesis, bracket, brace, tick), it adds the closer, so the patch works correctly.  However, when I delete the opener, it leaves the closer.  It would be nice, if the two are next to each other, then automatically delete the closer.  If some characters were typed in between, then I don't think the closer should be deleted.

wohlganger committed:
Changes made per csabella request.

' and " changed to quotedbl and quoteright.
Fixed keybindings appearing in config.
Comment style fix.
GetOption bool.
Code deduplication on p_close_event, p_open_event, t_open_event.
Needed to change test to use instantiated ParenClose object rather than class methods as a result of previous changes. Probably makes for a better test.

wohlganger commented:
Fixed your requests, working on the deleting openers/closers feature. Please wait to check, I've realized I've made a stupid mistake...

wohlganger committed:
fix stupid mistake

copied wrong file to wrong place between where I can test code and my git folder.

wohlganger commented:
Optional mutual deletion of openers / closers added.

wohlganger committed:
added mutual delete / backspace. minor fixes

wohlganger committed:
PEP8 fixes.

wohlganger committed:
replace type option of bool with 'bool'

idleConf expects a string bool, not type bool. whoops.

wohlganger committed:
Merge branch 'master' into 18875-Auto-insert-parens

wohlganger committed:
whitespace

wohlganger committed:
whitespace fixed with reindent.py

wohlganger committed:
trying to fix Tk for travis testing parenclose

wohlganger committed:
Set parenclose test as gui required

wohlganger committed:
Added option to distinguish between mutual delete and mutual backspace

In use, it was irritating for me to have both perform the action.  This way, users can pick one or both or none.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+"""
```
Add ParenClose to parenmatch.py.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+When you hit left paren or tick,
+automatically creates the closing paren or tick.
+
+Author: Charles M. Wohlganger
```
Remove.  This already has input from 3 other people.  I will list Charles as original author in the news entry.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+Add to the end of config-extensions.def :
+
+    [ParenClose]
+    enable = True
```
Delete.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+    paren_close = True
+    tick_close = True
+    skip_closures = True
+    mutual_delete = True
```
I think defaults should be 'off', as in no change from current behavior.  I am not sure what latter 2 mean.   See below.

terryjreedy replied:
Changing the specification of the new feature by adding action on deletion should have been discussed on the issue.  (This applies generally to changing specs.) Notepad++ does do anything special on deletion.  Binding the deletion keys here prevents binding them elsewhere, and I have an idea of where it might be needed, as opposed to optional.  I also think sometimes delete more and sometimes not is confusing.  The deletion part of 2f19ca8 and anything later should be reverted unless and until there is a clear agreement on adding this.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+    tick_close = True
+    skip_closures = True
+    mutual_delete = True
+    [ParenClose_cfgBindings]
```
Delete section.  Not needed for hard-coded features.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+    When loaded as an extension to IDLE, and paren_close is True, the symbols
+    ([{ will have their closures printed after them and the insertion cursor
+    moved between the two.  The same is true for tick closures and the symbols
+    ' and ".  When \'\'\' or """ are typed and tick_close is True, it will also
```
I will consider rewording docstring, but this is secondary.  Why are triples single quotes escaped?

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+                self.text.insert(pos, event.char * 3)
+                self.text.mark_set('insert', pos)
+            elif not self.p_close_event(event):
+                self.p_open_event(event)
```
I have not yet looked at the instance method code.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+                self.p_open_event(event)
+
+
+if __name__ == '__main__':
```
Already in parenmatch.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -63,3 +63,17 @@ z-text= Z
+tick_close = True
+skip_closures = True
+mutual_delete = True
+mutual_backspace = True
```
I think the configuration options should be discussed more on the issue tracker.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -63,3 +63,17 @@ z-text= Z
+t-open = <Key-quotedbl> <Key-quoteright>
+p-close = <Key-parenright> <Key-bracketright> <Key-braceright>
+back = <Key-BackSpace>
+delete = <Key-Delete>
```
Again, non-configurable bindings are not needed *here*.  On the other hand, the bindings will have to be added elsewhere.  Currently, these are done in EditorWindow.__init__.

In the future, I want to think about and maybe try putting feature bindings in the feature class.  Also, I believe they could be class rather than instance bindings, as they are the same for all instances.

EDIT: Key bindings could, maybe should, be done in a loop iterating through (key, event_handler) pairs, and in a separate method.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+            type='bool', default=True)
+        self.mutual_backspace = idleConf.GetOption(
+            'extensions', 'ParenClose', 'mutual_backspace',
+            type='bool', default=True)
```
Getting the options should be done in a `def reload(cls)`, the same as with ParenMatch.
`ParenClose.reload()` must be added after `ParenMatch.reload()` at the bottom of the file.
A call to `ParenClose.reload` must be added to configdialog wherever `ParenMatch.reload` is called, to make user changes take immediate effect.

terryjreedy commented on Lib/idlelib/idle_test/test_parenclose.py:
```diff
@@ -0,0 +1,97 @@
+            # this test. We are checking to make sure everything is
+            # correct before they would delete or backspace.
+
+            # Reset text and self.
```
`with self.subTest ...` is needed to make asserts independent.

terryjreedy commented on Lib/idlelib/idle_test/test_parenclose.py:
```diff
@@ -0,0 +1,97 @@
+        p_close = p.p_close_event
+        t_open = p.t_open_event
+        back = p.back_event
+        delete = p.delete_event
```
I do not like testing all 5 methods in one method.  On the other hand, I don't like repeating code, and using subtests allows multiple failures in one method.  A helper function might allow separate test functions without duplication, but this is a lower priority than getting good coverage to start with.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+    closers = {'(': ')', '[': ']', '{': '}', "'": "'", '"': '"'}
+
+    def __init__(self, editwin=None):
+        # Setting default to none makes testing easier.
```
I an not sure that this will work for a non-extension feature.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+Author: Charles M. Wohlganger
+        charles.wohlganger@gmail.com
+
+Last Updated: 13-Aug-2017 by Charles Wohlganger
```
Remove.  Already obsolete.  The git log has revision history.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+
+Last Updated: 13-Aug-2017 by Charles Wohlganger
+
+Add to the end of config-extensions.def :
```
Config options will be added to config-general.def.  The general tab already needs to be split for another issue.  This issue, and a complete patch, will depend on how we do that, and then how to lay out the option buttons.

terryjreedy commented on Lib/idlelib/ParenClose.py:
```diff
@@ -0,0 +1,115 @@
+    deleted before the new one is typed, effectively skipping over the closure.
+    If the cursor is between two closures and mutual_delete or mutual_backspace
+    is True, when the respective command is given, both closures are deleted.
+    '''
```
I don't really understand these.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -63,3 +63,17 @@ z-text= Z
 z-in= <Control-Shift-KeyRelease-Insert>
 [ZzDummy_bindings]
 z-out= <Control-Shift-KeyRelease-Delete>
+
```
As I said above, put in config-general.

terryjreedy commented on Lib/idlelib/idle_test/test_config.py:
```diff
@@ -439,21 +439,21 @@ def test_get_extensions(self):
+        eq(iGE(), ['ZzDummy', 'ParenClose'])
+        eq(iGE(editor_only=True), ['ZzDummy', 'ParenClose'])
+        eq(iGE(active_only=False), ['ZzDummy', 'ParenClose', 'DISABLE'])
+        eq(iGE(active_only=False, editor_only=True), ['ZzDummy', 'ParenClose', 'DISABLE'])
```
I believe that all these changes to test_config only apply to extensions, and are therefore not needed.

terryjreedy commented on Lib/idlelib/idle_test/test_parenclose.py:
```diff
@@ -0,0 +1,97 @@
+'''Test idlelib.parenclose.
```
Add MockEvent and ParenCloseTest to test_parenmatch.  I have not reviewed in detail, but have a couple of comments.

terryjreedy commented on Lib/idlelib/idle_test/test_parenclose.py:
```diff
@@ -0,0 +1,97 @@
+        delete = p.delete_event
+        for paren_close_set, tick_close_set, skip_closures_set, mutual_b_set, \
+            mutual_d_set, insert, func, pos, char, result, posend in [
+              (1, 1, 1, 1, 1, 'def abuse', p_open, '1.9', '(',
```
`for test_tuple in [<list of tuples>]` would allow test_tuple to be passed to a helper function.

terryjreedy reviewed (changes requested):
Now that we are no longer implementing new features as extensions, this patch needs to be rewritten before it can be properly review in action.  If someone else wants to do so before I get to it, please say so in a comment.

The comments below indicate the major changes needed.

terryjreedy committed:
Update ParenClose.py

Trigger a retest to see if CI produces same odd failures as on another PR.
